### ISSUE TITLE
Proposal to break out lib/controller into separate repo

### ISSUE THREAD

**Issue by bassam** [kubernetes-incubator/external-storage#423]
The Rook project uses the provisioner in lib/controller and also takes dependencies on kubernetes projects directly (apis, apimachinery, client-go etc.). Managing dependencies between Rook --> Kubernetes and External-Storage--> Kubernetes has been challenging. Furthermore, the External-Storage project seems to be lagging a few versions behind upstream Kubernetes, and to get everything updated it requires not only changes to lib/controller but also to all the storage projects hosted here.

We'd like to propose that we break out lib/controller into a separate project/repo that can be versioned closely with Kubernetes. The external-storage project would take a dependency (glide or dep) on the new libcontroller project (at a given commit hash).

This would enable users like Rook to have greater control of our versioning story and reduce the dependency issues. Please chime with your thoughts on this approach.

**Comment by wongma7** [comment]
Yes, this would be good.

If we were not a monorepo https://github.com/kubernetes-incubator/external-storage/pull/392 would have merged and the lib would have been ready for 1.8.0 pre-release.

The idea behind the monorepo has always been to allow for both: A) I update the repo with every k8s release or upon request, or B) somebody updates the repo themselves when they need it, e.g. @humblec and @kokhang. Note that I think this idea has worked fine until now, the difference this release is that for A) I've not found time to work on https://github.com/kubernetes-incubator/external-storage/pull/419 and B) nobody else can/wants to (besides @xingzhou who has already done most of the work) because the difficulty/tedium has increased with the repo size.

side note: the hypothetical separate repo must include also lib/leaderelection. the rest of lib/* is small/harmless/self-contained and should be included as well, as the stuff there is just non-essential helpers that a couple provisioners want to share e.g. gid allocator.

**Comment by bassam** [comment]
@wongma7 thanks! we'll get a PR going on this soon. who do we have to talk to get a new repo in the kuberenetes-incubator org?

**Comment by wongma7** [comment]
First we need the attention of @jsafrane / @childsb. After that the procedure I imagine will be something like when we  split off external-storage from nfs-provisioner, except this time we aren't abandoning the old repo. Don't recall exactly the steps, will need @childsb input first.

**Comment by wongma7** [comment]
BTW no need to get a PR, I'm almost done #419 .

**Comment by bassam** [comment]
@wongma7 I think it makes sense to break out `lib/` into a separate repo even after #419 is merged. Are you suggesting otherwise with your last comment?

**Comment by wongma7** [comment]
No, I still agree with you bout a separate repo.

**Comment by bassam** [comment]
Thanks @wongma7.

@childsb @saad-ali would you guys support this? If so I'd like to create a new repo in kubernetes-incubator for it. Looking at the process here https://github.com/kubernetes/community/blob/master/incubator.md I'm going to need a champion and send off an email to kubernetes-dev.

**Comment by jsafrane** [comment]
Isn't the library mature enough to have non-incubator repo? @childsb had a separate repo on his TODO list.

**Comment by wongma7** [comment]
According to https://github.com/kubernetes/community/blob/master/incubator.md#exiting-incubation we need the champion's approval (we have @jsafrane/@childsb) and sponsor's (@smarterclayton) approval to "graduate". However since we only want to "graduate" a part of the repo the situation is not typical.

@smarterclayton what do you think about graduating the external provisioner library portion of this repo? The goal of having a dedicated lib repo is to allow us to release the library in the most consumable way possible, and make it easier to maintain so we can more closely follow k8s master. It's becoming especially important since kube is no longer accepting new volume plugins and devs don't have much choice but to implement this pattern.

**Comment by childsb** [comment]
I've brought this up with @bgrant0607 and how we should proceed with the external-storage incubator, the provisioner, libraries and dependencies.. especially graduating them from incubation..

I have a kubernetes-storage ORG where we could start hosting this stuff, but its all TBD.. Lets keep this as an agenda item for kubecon and hopefully get direction from Brian there.

**Comment by bgrant0607** [comment]
(Yes, this is how far behind on email I am.)

In case you missed it, the new-ish repo policy:
https://github.com/kubernetes/community/blob/master/kubernetes-repositories.md
